distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso

```diff
@@ -793,32 +877,23 @@ type Input_Stream
     with_java_stream : (Java_Input_Stream -> Any) -> Any
     with_java_stream f = Managed_Resource.with this.stream_resource f
 
-## PRIVATE
-
-   Utility method for rewrapping Java exceptions into Enso panics.
-
-   Arguments:
-   - file: The file object.
-   - exception: The java exception that was encountered.
-rethrow_java : File -> Any -> Any
-rethrow_java file exception =
-    case exception of
-        Polyglot_Error exc ->
-            if Java.is_instance exc NoSuchFileException then
-                Panic.throw (No_Such_File_Error file)
-            if Java.is_instance exc AccessDeniedException then
-                Panic.throw (Access_Denied_Error file)
-            if Java.is_instance exc IOException then
-                Panic.throw (Io_Error exc.getMessage)
-            Panic.throw exception
-        _ -> Panic.throw exception
 
 ## PRIVATE
 
    Utility method for running an action with Java exceptions mapping.
 handle_java_exceptions file ~action =
+    rethrow_java exception =
+        result = case exception of
+            Polyglot_Error exc ->
+                if Java.is_instance exc NoSuchFileException then Error.throw (File_Not_Found file) else
+                    if Java.is_instance exc AccessDeniedException then Error.throw (Io_Error file ("You do not have permission to access the file")) else
+                        if Java.is_instance exc IOException then Error.throw (Io_Error file ("An IO error has occurred: " + exc.getMessage)) else
+                            Panic.throw exception
+            _ -> Panic.throw exception
+        result
+
     err = Panic.recover Any action
-    r = err.catch (here.rethrow_java file)
+    r = err.catch (rethrow_java)
```

Comment:
This is still incorrect. Now, a dataflow error coming from `action` will be 'promoted' to a Panic - that is also wrong, maybe even worse than making a panic into a dataflow error as previously (well, both can cause problems).

`Panic.recover` should only be used as a 'final' operation if we don't want to process the panic any further. `Panic.recover Any` should be used very rarely.  Please use `Panic.catch` which was added specifically for the purpose of avoiding mixing Panics with DEs.